lib/tree.rb

```diff
@@ -16,42 +16,131 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity && Space Complexity: O(h) where h is the height of the tree
+  # Tree height has a minimum of O(log n) levels. Worst case, an unbalanced tree height is O(n) levels (https://www.cs.cmu.edu/~adamchik/15-121/lectures/Trees/trees.html)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+
+    if @root.nil?
+      @root = new_node
+      return 
+    end
+
+    add_helper_method(@root, new_node)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper_method(current, new_node)
+    if new_node.key <= current.key
+      if current.left.nil?
+        return current.left = new_node
+      else
+        add_helper_method(current.left, new_node)
+      end
+    else
+      if current.right.nil?
+        return current.right = new_node
+      else
+        add_helper_method(current.right, new_node)
+      end
+    end 
```

Comment:
This can be simplified a bit

```suggestion
    return new_node if current.nil?
    
     if new_node.key <= current.key
       current.left = add_helper_method(current.left, new_node)
     else
       current.right = add_helper_method(current.right, new_node)
     end
       
      return current   
```